penrose-web/src/engine/EngineUtils.ts

```diff
@@ -306,3 +314,391 @@ export const makeTranslationDifferentiable = (trans: any): Translation => {
+    // TODO (error/warning): Warn if any of these properties are duplicated or do not exist in the shape constructor
+    gpi[decl.name.value] = { tag: "OptEval", contents: decl.value };
+  }
+
+  return gpi;
+};
+
+/**
+ * Insert an expression into the translation (mutating it), returning a reference to the mutated translation for convenience
+ * @param path path to a field or property
+ * @param expr new expression
+ * @param initTrans initial translation
+ *
+ */
+
+// TODO: Test this
+export const insertGPI = (path: Path, gpi: IFGPI<VarAD>, trans: Translation): Translation => {
+  let name, field;
+
+  switch (path.tag) {
+    case "FieldPath": {
+      [name, field] = [path.name, path.field];
+      // TODO: warning / error here
+      trans.trMap[name.contents.value][field.value] = gpi;
+      return trans;
+    }
+
+    default: {
+      throw Error("expected GPI");
+    }
+  }
+};
+
+// This function is a combination of `addField` and `addProperty` from `Style.hs`
+export const insertExpr = (path: Path, expr: TagExpr<VarAD>, initTrans: Translation): Translation => {
+  const trans = initTrans;
+  let name, field, prop;
+
+  // console.log("insertExpr: path, expr", path, expr);
+  // COMBAK: Deal with path aliasing for both field and property paths
+
+  switch (path.tag) {
+    case "FieldPath": {
+      [name, field] = [path.name, path.field];
+
+      // Initialize the field dict if it hasn't been initialized
+      if (!trans.trMap[name.contents.value]) {
+        trans.trMap[name.contents.value] = {};
+      }
+
+      let fexpr: FieldExpr<VarAD> = { tag: "FExpr", contents: expr };
+
+      // If it's a GPI, instantiate it (rule Line-Set-Ctor); otherwise put it in the translation as-is
+      if (expr.tag === "OptEval") {
+        const gpi: Expr = expr.contents;
+        if (gpi.tag === "GPIDecl") {
+          const [nm, decls]: [Identifier, PropertyDecl[]] = [gpi.shapeName, gpi.properties];
+
+          fexpr = {
+            tag: "FGPI",
+            contents: [nm.value, mkPropertyDict(decls)]
+          };
+        }
+      }
+
+      // For any non-GPI thing, just put it in the translation
+      // NOTE: this will overwrite existing expressions
+      // TODO: warning / error here
+      trans.trMap[name.contents.value][field.value] = fexpr;
+      return trans;
+    }
+
+    case "PropertyPath": {
+      [name, field, prop] = [path.name, path.field, path.property];
+
+      if (!trans.trMap[name.contents.value]) {
+        trans.trMap[name.contents.value] = {};
+      }
+
+      const fieldRes: FieldExpr<IVarAD> = trans.trMap[name.contents.value][field.value];
+
+      if (fieldRes.tag === "FExpr") {
+        // Deal with GPI aliasing (i.e. only happens if a GPI is aliased to another, and some operation is performed on the aliased GPI's property, it happens to the original)
+        // TODO: Test this
+        if (fieldRes.contents.tag === "OptEval") {
+          if (fieldRes.contents.contents.tag === "FieldPath") {
```

Comment:
Do you think the types are nested too much? If so, we can refactor them into simpler ones wherever applicable.